### PR TITLE
[Media]Implemented dynamic Gpu session priority

### PR DESCRIPTION
**Pull request by dylanintel**
[Media]Implemented dynamic Gpu session priority
Implemented the media and vp Gpu context priority setting.

Signed-off-by: Wang Dylan <dylan.wang@intel.com>

### CODE REVIEW

File: media_driver/linux/common/ddi/media_libva_common.cpp [hunk 1]
```diff
@@ -352,3 +353,86 @@ void* DdiMedia_GetContextFromVABufferID (PDDI_MEDIA_CONTEXT mediaCtx, VABufferID
 
     return ctx;
 }
+
+int32_t DdiMedia_GetGpuPriority(
+    VADriverContextP     ctx,
+    VABufferID           *buffers,
+    int32_t              numBuffers,
+    bool                 *bPriorityUpdate,
```
Comment (Xiaogangli-intel) [inline media_driver/linux/common/ddi/media_libva_common.cpp:361]: Please follow/align the naming rules of variables.

File: media_driver/linux/common/ddi/media_libva_common.h [hunk 2]
```diff
@@ -679,4 +679,42 @@ void* DdiMedia_GetContextFromVABufferID (PDDI_MEDIA_CONTEXT mediaCtx, VABufferID
...
+//!     update priority
+//! \param  [out] priority
+//!     priority
+//! \return     int32_t
+//!
+int32_t DdiMedia_GetGpuPriority(
+    VADriverContextP     ctx,
+    VABufferID           *buffers,
+    int32_t              numBuffers,
+    bool                 *bPriorityUpdate,
```
Comment (Xiaogangli-intel) [inline media_driver/linux/common/ddi/media_libva_common.h:701]: Please follow/align the naming rules of variables.

**Review by Xiaogangli-intel — CHANGES_REQUESTED** [review]

File: media_driver/linux/common/os/i915_production/mos_bufmgr.c [hunk 3]
```diff
@@ -2957,7 +2957,7 @@ do_exec2(struct mos_linux_bo *bo, int used, struct mos_linux_context *ctx,
     case I915_EXEC_DEFAULT:
         break;
     }
-
+    printf("============do_exec2 on context id %d==========\n", ctx->ctx_id);
```
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/linux/common/os/i915_production/mos_bufmgr.c:2960]: yeah, I remain them and I will remove all. All of changes will be resubmitted later.

**Review by Xiaogangli-intel — CHANGES_REQUESTED** [review]

File: media_driver/linux/common/ddi/media_libva_common.cpp [hunk 4]
```diff
@@ -352,3 +353,75 @@ void* DdiMedia_GetContextFromVABufferID (PDDI_MEDIA_CONTEXT mediaCtx, VABufferID
...
+        DDI_CHK_NULL(buf, "Invalid buffer.", VA_STATUS_ERROR_INVALID_BUFFER);
+
+        if((int32_t)buf->uiType == VAContextParameterAdjustBufferType)
+        {
+            //Read the priority from the VAContextParameterAdjustBuffer
+            DdiMedia_MapBuffer(ctx, buffers[i], &data);
+            DDI_CHK_NULL(data, "nullptr data.", VA_STATUS_ERROR_INVALID_BUFFER);
+
+            VAContextParameterAdjustBuffer *ContextParamBuf = (VAContextParameterAdjustBuffer *)data;
+            DDI_CHK_NULL(ContextParamBuf, "nullptr ContextParamBuf.", VA_STATUS_ERROR_INVALID_BUFFER);
```
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/linux/common/ddi/media_libva_common.cpp]: Done. Thanks.

File: media_driver/linux/common/ddi/media_libva_common.cpp [hunk 5]
```diff
@@ -352,3 +353,75 @@ void* DdiMedia_GetContextFromVABufferID (PDDI_MEDIA_CONTEXT mediaCtx, VABufferID
...
+    for (int32_t i = 0; i < numBuffers; i++)
+    {
+        DDI_MEDIA_BUFFER *buf = DdiMedia_GetBufferFromVABufferID(mediaCtx, buffers[i]);
+        DDI_CHK_NULL(buf, "Invalid buffer.", VA_STATUS_ERROR_INVALID_BUFFER);
+
+        if((int32_t)buf->uiType == VAContextParameterAdjustBufferType)
+        {
+            //Read the priority from the VAContextParameterAdjustBuffer
+            DdiMedia_MapBuffer(ctx, buffers[i], &data);
+            DDI_CHK_NULL(data, "nullptr data.", VA_STATUS_ERROR_INVALID_BUFFER);
```
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/linux/common/ddi/media_libva_common.cpp]: Done.

File: media_driver/linux/common/ddi/media_libva_common.cpp [hunk 6]
```diff
@@ -352,3 +353,75 @@ void* DdiMedia_GetContextFromVABufferID (PDDI_MEDIA_CONTEXT mediaCtx, VABufferID
...
+{
+    void *        data;
+    uint32_t      updateSessionPriority  = 0;
+    uint32_t      priorityValue          = 0;
+    int32_t       priorityIndexInBuf     = -1;
+
+    DDI_CHK_NULL(ctx,                "nullptr context in DdiDecode_SetGpuPriority!", VA_STATUS_ERROR_INVALID_CONTEXT);
+
+    PDDI_MEDIA_CONTEXT mediaCtx = DdiMedia_GetMediaContext(ctx);
+    DDI_CHK_NULL(mediaCtx, "nullptr mediaCtx", VA_STATUS_ERROR_INVALID_CONTEXT);
```
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/linux/common/ddi/media_libva_common.cpp]: Done.

File: media_driver/linux/common/codec/ddi/media_libva_decoder.cpp [hunk 7]
```diff
@@ -603,3 +623,35 @@ VAStatus DdiDecode_DestroyContext (
...
+        PMOS_INTERFACE osInterface = decCtx->pCodecHal->GetOsInterface();
+        DDI_CHK_NULL(osInterface, "nullptr osInterface.", VA_STATUS_ERROR_ALLOCATION_FAILED);
+        osInterface->pfnSetGpuPriority(osInterface, priority);
+    }
+#ifdef _MMC_SUPPORTED
+    //set the priority for decomp interface
+    if(mediaCtx->pMediaMemDecompState)
+    {
+        MediaVeboxDecompState *mediaVeboxDecompState = static_cast<MediaVeboxDecompState*>(mediaCtx->pMediaMemDecompState);
+        if(mediaVeboxDecompState)
```
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/linux/common/codec/ddi/media_libva_decoder.cpp]: Done.

File: media_driver/linux/common/os/i915_production/mos_bufmgr.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/linux/common/os/i915_production/mos_bufmgr.c]: Done.

File: media_driver/linux/common/os/mos_interface.cpp [hunk 8]
```diff
@@ -2267,6 +2270,58 @@ GMM_CLIENT_CONTEXT *MosInterface::GetGmmClientContext(
...
+
+    PMOS_OS_CONTEXT pOsContext = (PMOS_OS_CONTEXT)streamState->perStreamParameters;
+    if (pOsContext == nullptr)
+    {
+        MOS_OS_ASSERTMESSAGE("Failed to set the gpu priority");
+        return;
+    }
+
+    int32_t ret = mos_set_context_param(pOsContext->intel_context, 0, I915_CONTEXT_PARAM_PRIORITY,(uint64_t)priority);
+    printf("===========Set the priority %d for the context id %d==============\n", priority, pOsContext->intel_context->ctx_id);
```
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/linux/common/os/mos_interface.cpp]: Done.

File: media_driver/media_driver_next/linux/common/os/mos_gpucontext_specific_next.cpp [hunk 9]
```diff
@@ -1391,6 +1391,29 @@ void GpuContextSpecificNext::IncrementGpuStatusTag()
...
+    {
+        return;
+    }
+
+    for (int32_t i=0; i<MAX_ENGINE_INSTANCE_NUM+1; i++)
+    {
+        if (m_i915Context[i] != nullptr)
+        {
+            int32_t ret = mos_set_context_param(m_i915Context[i], 0, I915_CONTEXT_PARAM_PRIORITY,(uint64_t)priority);
+            printf("===============Set the priority %d for context %d==============\n", priority, m_i915Context[i]->ctx_id);
```
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/media_driver_next/linux/common/os/mos_gpucontext_specific_next.cpp]: Done.

File: media_driver/linux/common/codec/ddi/media_libva_decoder.cpp [hunk 10]
```diff
@@ -171,6 +172,12 @@ VAStatus DdiDecode_RenderPicture (
     int32_t             numBuffers
 )
 {
+    VAStatus        va                      = VA_STATUS_SUCCESS;
+    int32_t         priority                = 0;
+    int32_t         numOfBuffers            = numBuffers;
+    int32_t         priorityIndexInBuffers  = -1;
+    bool            bUpdatePriority         = false;
```
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/linux/common/codec/ddi/media_libva_decoder.cpp]: Done.

File: media_driver/linux/common/codec/ddi/media_libva_encoder.cpp [hunk 11]
```diff
@@ -579,6 +580,12 @@ VAStatus DdiEncode_RenderPicture(
     VABufferID      *buffers,
     int32_t          num_buffers)
 {
+    VAStatus        vaStatus                  = VA_STATUS_SUCCESS;
+    int32_t         priority                  = 0;
+    int32_t         numOfBuffers              = num_buffers;
+    int32_t         priorityIndexInBuffers    = -1;
+    bool            bUpdatePriority           = false;
```
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/linux/common/codec/ddi/media_libva_encoder.cpp]: Done.

File: media_driver/linux/common/vp/ddi/media_libva_vp.c [hunk 12]
```diff
@@ -3556,6 +3557,10 @@ VAStatus DdiVp_RenderPicture (
     void                      *pData;
     uint32_t                  ctxType;
     VAStatus                  vaStatus;
+    int32_t                   priority                  = 0;
+    int32_t                   numOfBuffers              = num_buffers;
+    int32_t                   priorityIndexInBuffers    = -1;
+    bool                      bUpdatePriority           = false;
```
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/linux/common/vp/ddi/media_libva_vp.c]: Done.

File: media_driver/agnostic/common/os/mos_os.h [hunk 13]
```diff
@@ -1263,6 +1264,29 @@ typedef struct _MOS_INTERFACE
...
+    //!
+    //! \brief    Get Gpu Priority
+    //! \param    [in] pOsInterface
+    //!           pointer to the current gpu context.
+    //! \param    [out] pPriority
+    //!           pointer to the priority get from gpu context.
+    //!
+    void (*pfnGetGpuPriority)(
+          PMOS_INTERFACE          pOsInterface,
+          int32_t*                pPriority);
```
[the remark this answers is not in the document]
Comment (dylanintel) [reply media_driver/agnostic/common/os/mos_os.h]: Done.

File: media_driver/linux/common/ddi/media_libva_common.cpp (hunk 1 above)
Comment (dylanintel) [reply media_driver/linux/common/ddi/media_libva_common.cpp]: Done.

File: media_driver/linux/common/ddi/media_libva_common.h (hunk 2 above)
Comment (dylanintel) [reply media_driver/linux/common/ddi/media_libva_common.h]: Done.

**Review by Xiaogangli-intel — APPROVED** [review]

**Pull request merged**